### ISSUE TITLE
ChannelClosedException due to spurious termination of running containers

### ISSUE THREAD

**Issue by StephanErb** [jenkinsci/docker-plugin#653]
#### Setup
 - docker-plugin version: `1.1.4`
 - jenkins version: `2.107.3`(LTS) but also confirmed with recent weeklys as of May 2018
 - docker engine version: `18.02.0~ce-0~debian`
 - Setup consists of a Jenkins master with one additional host running Docker.  Agents are started via the ssh-agent connect method.

#### Symptom
A low percentage of our Jenkins jobs end up hanging. Affected executors are shown as `docker-5d5fedb1d75d0  (offline)  (suspended)` and the job itself contains the output:
```
Cannot contact docker-5d5fedb1d75d0:
    java.io.IOException:
        remote file operation failed: /home/jenkins/workspace/x/y/z/systemtests@2/test_definitions at hudson.remoting.Channel@1ad4d27c:docker-5d5fedb1d75d0:
        hudson.remoting.ChannelClosedException: 
            Channel "unknown": Remote call on docker-5d5fedb1d75d0 failed. The channel is closing down or has closed down
```

Jenkins master logs indicate that _something_ decided to terminate running containers. Sometimes this happens after a few minutes, but sometimes only after 1-2 hours for long running jobs.
```
May 22 20:57:05 s5 java[29653]: INFO: Trying to run container for node 5d5fedb1d75d0 from image: images-build.blue-yonder.org:5000/by/debian_8_jenkins:stable
May 22 20:57:06 s5 java[29653]: INFO: Started container ID 99356f7db1185054da6d62f18f543eb9213e92c950881a1d1e1794115c57dc6d for node 5d5fedb1d75d0 from image: images-build.blue-yonder.org:5000/by/debian_8_jenkins:stable
May 22 20:57:11 s5 java[29653]: [05/22/18 20:57:11] SSH Launch of docker-5d5fedb1d75d0 on 10.1.32.51 completed in 2,523 ms
May 22 21:08:19 s5 java[29653]: INFO: Disconnected computer for slave 'docker-5d5fedb1d75d0'.
May 22 21:08:19 s5 java[29653]: INFO: Removed Node for slave 'docker-5d5fedb1d75d0'.
May 22 21:08:21 s5 java[29653]: INFO: Stopped container '99356f7db1185054da6d62f18f543eb9213e92c950881a1d1e1794115c57dc6d' for slave 'docker-5d5fedb1d75d0'.
May 22 21:08:23 s5 java[29653]: INFO: Removed container '99356f7db1185054da6d62f18f543eb9213e92c950881a1d1e1794115c57dc6d' for slave 'docker-5d5fedb1d75d0'.
May 22 21:09:16 s5 java[29653]: INFO: Attempting to reconnect docker-5d5fedb1d75d0
May 22 21:09:16 s5 java[29653]: com.github.dockerjava.api.exception.NotFoundException: {"message":"No such container: 99356f7db1185054da6d62f18f543eb9213e92c950881a1d1e1794115c57dc6d"}
May 22 21:11:16 s5 java[29653]: INFO: Attempting to reconnect docker-5d5fedb1d75d0
May 22 21:11:16 s5 java[29653]: com.github.dockerjava.api.exception.NotFoundException: {"message":"No such container: 99356f7db1185054da6d62f18f543eb9213e92c950881a1d1e1794115c57dc6d"}
May 22 21:13:16 s5 java[29653]: INFO: Attempting to reconnect docker-5d5fedb1d75d0
May 22 21:13:16 s5 java[29653]: com.github.dockerjava.api.exception.NotFoundException: {"message":"No such container: 99356f7db1185054da6d62f18f543eb9213e92c950881a1d1e1794115c57dc6d"}
May 22 21:15:16 s5 java[29653]: INFO: Attempting to reconnect docker-5d5fedb1d75d0
May 22 21:15:16 s5 java[29653]: com.github.dockerjava.api.exception.NotFoundException: {"message":"No such container: 99356f7db1185054da6d62f18f543eb9213e92c950881a1d1e1794115c57dc6d"}
May 22 21:15:59 s5 java[29653]: WARNING: Caught exception evaluating: it.containerId in /computer/docker-5d5fedb1d75d0/. Reason: java.lang.reflect.InvocationTargetException
May 22 21:15:59 s5 java[29653]: WARNING: Caught exception evaluating: it.containerId in /computer/docker-5d5fedb1d75d0/. Reason: java.lang.reflect.InvocationTargetException
May 22 21:15:59 s5 java[29653]: WARNING: Caught exception evaluating: it.containerId in /computer/docker-5d5fedb1d75d0/. Reason: java.lang.reflect.InvocationTargetException
May 22 21:15:59 s5 java[29653]: WARNING: Caught exception evaluating: it.containerId in /computer/docker-5d5fedb1d75d0/. Reason: java.lang.reflect.InvocationTargetException
May 22 21:15:59 s5 java[29653]: WARNING: Caught exception evaluating: it.containerId in /computer/docker-5d5fedb1d75d0/. Reason: java.lang.reflect.InvocationTargetException
May 22 21:17:16 s5 java[29653]: INFO: Attempting to reconnect docker-5d5fedb1d75d0
May 22 21:17:16 s5 java[29653]: com.github.dockerjava.api.exception.NotFoundException: {"message":"No such container: 99356f7db1185054da6d62f18f543eb9213e92c950881a1d1e1794115c57dc6d"}
```

The interesting bit is that `Disconnected computer for slave 'docker-5d5fedb1d75d0'.` The context of this log message looks like this:

```
May 22 21:08:19 s5 java[29653]: May 22, 2018 9:08:19 PM hudson.model.Run execute
May 22 21:08:19 s5 java[29653]: INFO: admin/test_labels #657 main build action completed: SUCCESS
May 22 21:08:19 s5 java[29653]: May 22, 2018 9:08:19 PM io.jenkins.docker.DockerTransientNode$1 println
May 22 21:08:19 s5 java[29653]: INFO: Disconnected computer for slave 'docker-5d5fedb1d75d0'.
May 22 21:08:19 s5 java[29653]: May 22, 2018 9:08:19 PM io.jenkins.docker.DockerTransientNode$1 println
May 22 21:08:19 s5 java[29653]: INFO: Removed Node for slave 'docker-5d5fedb1d75d0'.
May 22 21:08:19 s5 java[29653]: May 22, 2018 9:08:19 PM io.jenkins.docker.DockerTransientNode$1 println
May 22 21:08:19 s5 java[29653]: INFO: Removed container 'f6f330ff09db37fcc9e5dfb813634c0d5276044baec73fc4affadbe9e2d2d898' for slave 'docker-5d69904a90f09'.
May 22 21:08:19 s5 java[29653]: May 22, 2018 9:08:19 PM io.jenkins.docker.DockerTransientNode$1 println
May 22 21:08:19 s5 java[29653]: INFO: Stopped container '827e4ffa9e75253adfe4f5eb83d428f2e4b2602564861cb17a9b299c7b22a9cd' for slave 'docker-5d698fd47871d'.
May 22 21:08:21 s5 java[29653]: May 22, 2018 9:08:21 PM io.jenkins.docker.DockerTransientNode$1 println
May 22 21:08:21 s5 java[29653]: INFO: Stopped container '99356f7db1185054da6d62f18f543eb9213e92c950881a1d1e1794115c57dc6d' for slave 'docker-5d5fedb1d75d0'.
```

So what seems to happen that a unrelated job finishes but then our container gets terminated/disconnected. This _could_ be accidental due to ordering of logs, but the behaviour seems to show up in all examples I have looked: **A job finishes and then seemingly the container/agent of another job gets terminated.**

The source of the log messages seem to indicate that the docker plugin is involved here. Is there a chance for a race condition that could lead to the describe behaviour? 

Thanks for your input!

**Comment by StephanErb** [comment]
Thanks a lot for your quick reply! 

> Question: Are your jobs pipeline jobs or old-style projects?
Jenkins allows for "ephemeral" tasks to run on a slave node, which does not occupy an executor and does not affect the slave node's lifecycle, so if Jenkins was executed some ephemeral activities on a container that was busy running a build, and then the build finishes, the container would be terminated without regard to the ephemeral activities.
e.g. I'm pretty sure there's code in the Jenkins vSphere plugin to ensure that ephemeral stuff doesn't run on vSphere slaves that might be disposed of at any point - maybe what you're seeing indicates that the docker-plugin needs the same stuff?

Very good catch! I was able to find some logs that support this theory.

Failure in the aborted `myjob` run on container `docker-64b5b8e696bd5`:
```
08:47:52 Running on docker-64b5b8e696bd5 on build-agent-01 in /home/jenkins/workspace/myjob
....
....
09:08:19 [evaluation tests] Cannot contact docker-64b5b8e696bd5: 
    java.io.IOException:
        remote file operation failed: /home/jenkins/workspace/myjob at hudson.remoting.Channel@15a42e59:docker-64b5b8e696bd5:
            hudson.remoting.ChannelClosedException: 
                Channel "unknown": Remote call on docker-64b5b8e696bd5 failed. The channel is closing down or has closed down
```
 
In the Jenkins logs from the UI, I see the termination message for that particular container:
```
May 24, 2018 9:08:18 AM com.nirima.jenkins.plugins.docker.strategy.DockerOnceRetentionStrategy done
FINE: terminating docker-64b5b8e696bd5 since admin/test_labels #693 seems to be finished
```
Interesting fact here is that  `admin/test_labels` is a not equal to the terminated `myjob` above. It is a multi-configuration job that just happens to run simultaneously.

Output of the job confirms that they are indeed running on the same container:
```
09:08:00 Started by timer
09:08:00 [EnvInject] - Loading node environment variables.
09:08:00 Building remotely on docker-64b5b8e696bd5 on build-agent-01 (debian_9_computenode_testing Debian_9_testing) in workspace /home/jenkins/workspace/admin/test_labels
09:08:00 Triggering Administration » test_labels » Debian_8
09:08:00 Triggering Administration » test_labels » Debian_9
09:08:10 Administration » test_labels » Debian_8 completed with result SUCCESS
09:08:18 Administration » test_labels » Debian_9 completed with result SUCCESS
09:08:18 Finished: SUCCESS
```

So all in all this seems to match the symptoms of  https://issues.jenkins-ci.org/browse/JENKINS-35566 but is made worse by the fact that the termination of the "ephemeral master build" takes down a container that is already busy running something else.

**Comment by pjdarton** [comment]
Multi-configuration?  You should have mentioned that before - multi-configuration jobs are a bit counter-intuitive (and this bit us in the backside too):
-  You need to tell your multi-configuration job to "restrict where this build can run" and tell it to only run on the master.
 - The "do the actual build steps" bits will run on whatever nodes are available or dictated by your axis configuration.
Note that this does NOT mean that it'll run stuff on the master - it won't run anything on the master - but the "overall coordinating job" will show up as running on the master node (but doesn't require an executor as it's a flyweight task).

However, it does sounds like the docker plugin could do with some "prevent flyweight tasks from running here" logic.  Maybe.  This isn't an aspect of Jenkins and/or cloud plugins that I fully understand hence I can't state with confidence that adding such a bit of code would be free of unintended consequences - it might fix this issue but cause worse ones.
FYI the core Jenkins folks aren't too keen on multi-configuration builds - I'm told that all the cool kids are using pipelines these days and, personally, I've found that pipelines give better control/functionality than a multi-configuration build does ... but it is (of course) "yet another technology to master".  Where I work, we've got a mix of old multi-configuration builds and new pipeline stuff and I suspect that this will continue to be the case for some time to come.

**Comment by StephanErb** [comment]
Thanks a lot.

Fortunately we have already migrated roughly 95% of all jobs to pipelines and can hopefully either eliminate the remaining multi-configuration jobs or at least restrict the coordinator/flyweight to just run on master.

**Comment by pjdarton** [comment]
@magn2 As stated in my [earlier comment](https://github.com/jenkinsci/docker-plugin/issues/653#issuecomment-391676366), if you're using multi-configuration jobs, you need to ensure that the coordinator job (the one that triggers all the other bits of the multi-configuration job and monitors them for completion) is restricted to only run on the master.
If you have Jenkins running "flyweight jobs" (e.g. the mult-configuration's coordinator job) on a docker node then, when it ends, it'll shut down the container (as they're intended to be one-use items) and "flyweight jobs" don't count against the executor count so they neither count as an ongoing job (to keep the container alive) or as a job using the one-and-only executor (to keep other jobs from using their container).

TL;DR: Keep "flyweight jobs" locked to just the master (even if you have the number of executors set to zero on the master).

**Comment by magn2** [comment]
Sorry, I did not make that clear: I have no multi-configuration jobs, only ordinary pipelines.
Is there a coordinator also in this case?
And why would it make any difference that it works on instance cap 1?

**Comment by pjdarton** [comment]
Whether there's a "coordinator" job or not will depend entirely on the code you've got in your pipelines.
It's possible to write a pipeline such that it runs somewhere as a "flyweight" activity, but it is more normal to write it as a normal activity that will "consume" an executor.
i.e. it is _very_ dependent on your pipeline code.

To work out what's going on, one would need to dig through the logs, so the first step would be to collect some detailed logs covering the period when you see this.
If you can obtain logs from Jenkins, including the docker-plugin at level "ALL", so we can see exact timestamps of build activity starting/ending and the docker plugin starting/terminating containers and/or slave nodes then we'll be able to build up a picture of what's happening, where it's happening, and what's going wrong.
(and if you do this, please do it with the latest docker-plugin, as debugging older versions is more difficult)

**Comment by pjdarton** [comment]
FAO everyone who's commented here:
There's a new version of the plugin released.  I'd appreciate it if you could update to that and see if you can reproduce this issue and, if so, gather up all the logs generated at the time too (as well as details of your setup, templates etc)
Without those logs, we don't stand much of a chance figuring out what's causing this.

**Comment by willis7** [comment]
I was hitting the same issue and ended up at this page. After some debugging I found that the problem was caused by the process being run in the container.

For those curious to know the issue was being caused by the scala build tool on a multi-project build.

**Comment by willis7** [comment]
My last message was pretty rubbish - and fairly inaccurate. After some investigation I found an error code 137 after calling `docker inspect` on a terminated container. The docker docs suggest this is similar to a `kill -9` which hinted towards an external actor killing the container.

My colleague spotted that a flag in Jenkins may be the culprit.

`Jenkins -> Manage Jenkins -> Manage Nodes -> Configure -> Response Time`

 This monitors the round trip network response time from the master to the agent, and if it goes above a threshold repeatedly, it marks the agent offline.This is useful for detecting unresponsive agents, or other network problems that clog the communication channel. More specifically, the master sends a no-op command to the agent, and checks the time it takes to get back the result of this no-op command.

Once I unchecked this flag, we got consistent build passes.

**Comment by ajorpheus** [comment]
To supplement @willis7's post, what you should look for is the following snippet in the `docker inspect` log for the container(s):

```
 "State": {
            "Status": "exited",
			.....
            "OOMKilled": false,
			.....
            "ExitCode": 137,
            "Error": "",
 }
```

Note that the ExitCode of 137 is clearly documented [here](https://success.docker.com/article/what-causes-a-container-to-exit-with-code-137) as being equivalent to a `kill -9`, suggesting that the container is being killed.  Additionally, note that an ExitCode 137 may also be triggered by an out of memory error, so watch out for the value of `OOMKilled` in the `State` of the killed container.

To rule out a co-incidence, the "Response Time" was enabled and the builds began failing intermittently with the ChannelCloseException again.

"Response Time"
![2018-09-27_13-08-20 1](https://user-images.githubusercontent.com/1554671/46145066-ce5d5800-c256-11e8-9fab-a1019a8c523e.png)
